src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
             SetResult(node, GetReturnTypeWithState(method), method.ReturnTypeWithAnnotations);
         }
 
+        private bool IsStaticEqualsMethod(MethodSymbol method)
+        {
+            if (method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__EqualsObjectObject)))
+            {
+                return true;
+            }
+            else if (method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__ReferenceEquals)))
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private bool IsObjectEqualsMethodOrOverride(MethodSymbol method)
+        {
+            if (method.ParameterCount != 1
+                || method.ReturnType.SpecialType != SpecialType.System_Boolean
+                || method.Name != WellKnownMemberNames.ObjectEquals)
+            {
+                return false;
+            }
+
+            var objectEqualsMethod = compilation.GetSpecialTypeMember(SpecialMember.System_Object__Equals);
+            return method.GetLeastOverriddenMethod(accessingTypeOpt: null).Equals(objectEqualsMethod);
+        }
+
+
+        private bool IsIEquatableEqualsMethodOrImplementation(MethodSymbol method)
+        {
+            if (method.ParameterCount != 1
+                || method.MethodKind != MethodKind.Ordinary
+                || method.ReturnType.SpecialType != SpecialType.System_Boolean
+                || method.Name != WellKnownMemberNames.ObjectEquals)
+            {
+                return false;
+            }
+
+            var iEquatableType = compilation.GetWellKnownType(WellKnownType.System_IEquatable_T);
+            if (iEquatableType is null)
+            {
+                return false;
+            }
+
+            var parameterType = method.Parameters[0].TypeWithAnnotations;
+            var constructedIEquatableType = iEquatableType.Construct(ImmutableArray.Create(parameterType));
+            var members = constructedIEquatableType.GetMembers(WellKnownMemberNames.ObjectEquals);
+            if (members.Length != 1)
+            {
+                // If the Equals method is missing on the interface definition, or it's overloaded, give up.
+                return false;
+            }
+
+            var constructedIEquatableEqualsMethod = members[0];
+            if (constructedIEquatableEqualsMethod.Equals(method))
+            {
+                return true;
```

Comment:
>return true; [](start = 16, length = 12)

This method has two code paths that could return true. I didn't understand why we need that. It seems like the path that uses `FindImplementationForInterfaceMember` would always work, no?

Reply:
I found `m1.FindImplementationForInterfaceMember(m2)` returned null when m1 was equal to m2. I can double check on it.

---
In reply to: [297383790](https://github.com/dotnet/roslyn/pull/36722#discussion_r297383790) [](ancestors = 297383790)

Reply:
I double checked on this, removing this return breaks the following tests
- IEqualityComparerEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue
- IEqualityComparer_SubInterfaceEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue
- IEqualityComparer_SubSubInterfaceEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue
- IEquatableEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue

---
In reply to: [297392516](https://github.com/dotnet/roslyn/pull/36722#discussion_r297392516) [](ancestors = 297392516,297383790)